lib/logstash/inputs/s3/remote_file.rb

```diff
@@ -0,0 +1,90 @@
+# encoding: utf-8
+require "logstash/inputs/base"
+
+require "logstash/inputs/s3/stream_downloader"
+require "forwardable"
+
+module LogStash module Inputs class S3 < LogStash::Inputs::Base
+  class RemoteFile
+    class NoKeepAlive
+      def self.notify!
+      end
+
+      def self.complete!
+      end
+    end
+
+    GZIP_EXTENSION = ".gz"
+    FILE_MODE = "w+b"
+
+    extend Forwardable
+
+    attr_reader :remote_object, :metadata, :file
+    attr_accessor :download_to_path
+
+    def_delegators :@remote_object, :key, :content_length, :last_modified, :etag, :bucket_name
+
+    def initialize(object, keep_alive = NoKeepAlive)
+      @remote_object = object
+      @keep_alive = keep_alive
+      @downloaded = false
+      download_to_path = Dir.tmpdir
+    end
+
+    def download!
+      @file = StreamDownloader.fetcher(self).fetch
+      @downloaded = true
+    end
+
+    def download_to
+      # Lazy create FD
+      @download_to ||= begin
+                         FileUtils.mkdir_p(download_to_path)
```

Comment:
It seems like this doesn't work if the `key` is a full path instead of a single file, so maybe this should read more like...
```ruby
full_local_path = ::File.join(download_to_path, key)
FileUtils.mkdir_p(::File.dirname(full_local_path))
::File.open(full_local_path, FILE_MODE)
```